deepinv/transform/base.py

```diff
@@ -1,13 +1,80 @@
 from __future__ import annotations
 import torch
+from typing import Tuple, Callable, Any
+
+
+class Param(torch.Tensor):
+    """
+    Helper class that stores a tensor parameter for the sole purpose of allowing overriding negation.
+    """
+
+    @staticmethod
+    def __new__(cls, x, neg=None):
+        return torch.Tensor._make_subclass(cls, x)
+
+    def __init__(self, x, neg: Callable = lambda x: -x):
+        self._neg = neg
+
+    def __neg__(self):
+        return self._neg(torch.Tensor._make_subclass(torch.Tensor, self))
 
 
 class Transform(torch.nn.Module):
     """Base class for image transforms.
 
-    Multiply transforms to create compound transforms (direct product of groups).
+    The base transform implements transform arithmetic and other methods to invert transforms and symmetrize functions.
+
+    All transforms must implement ``get_params()`` to randomly generate e.g. rotation degrees or shift pixels, and ``transform()`` to deterministically transform an image given the params.
+
+    To implement a new transform, please reimplement ``get_params()`` and ``transform()`` (with a ``**kwargs`` argument). See respective methods for details.
+
+    Also handle deterministic (non-random) transformations by passing in fixed parameter values.
+
+    |sep|
+
+    Examples:
+
+        Randomly transform an image:
+
+        >>> import torch
+        >>> from deepinv.transform import Shift, Rotate
+        >>> x = torch.rand((1, 1, 2, 2)) # Define random image (B,C,H,W)
+        >>> transform = Shift() # Define random shift transform
+        >>> transform(x).shape
+        torch.Size([1, 1, 2, 2])
+
+        Deterministically transform an image:
+
+        >>> y = transform(transform(x, x_shift=[1]), x_shift=[-1])
+        >>> torch.all(x == y)
+        tensor(True)
+
+        Multiply transforms to create compound transforms (direct product of groups) - similar to ``torchvision.transforms.Compose``:
+
+        >>> rotoshift = Rotate() * Shift() # Chain rotate and shift transforms
+        >>> rotoshift(x).shape
+        torch.Size([1, 1, 2, 2])
+
+        Sum transforms to create stacks of transformed images (along the batch dimension).
+
+        >>> transform = Rotate() + Shift() # Stack rotate and shift transforms
+        >>> transform(x).shape
+        torch.Size([2, 1, 2, 2])
+
+        Randomly select from transforms - similar to ``torchvision.transforms.RandomApply``:
+
+        >>> transform = Rotate() | Shift() # Randomly select rotate or shift transforms
+        >>> transform(x).shape
+        torch.Size([1, 1, 2, 2])
+
+        Symmetrize a function for Reynolds Averaging:
+
+        >>> f = lambda x: x.pow(2) # Function to be symmetrized
```

Comment:
This function is already rotoshift symmetric...

what about
```suggestion
        >>> f = lambda x: x[..., 0]*x # Function to be symmetrized
```

Reply:
I'm also a bit unsure of whether symmetrize would work well with composed transforms like rotoshift?